### DISCUSSION TITLE
help bundle install error

### DISCUSSION THREAD

**Discussion by pandoli365** [mastodon/mastodon#30458] [Q&A]
When I build the bundle, the following error occurs. How can I resolve this error?

The OS is using arch

The error occurs after updating the system package.
If there is anyone who is familiar with Arch, does anyone know if there is a conflicting package in the update before the 22nd?

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/cbor-0.5.9.6/ext/cbor
/var/lib/mastodon/.rbenv/versions/3.2.3/bin/ruby extconf.rb
checking for ruby/st.h... yes
checking for st.h... no
checking for rb_str_replace() in ruby.h... yes
checking for rb_big_new() in ruby.h... yes
checking for rb_intern_str() in ruby.h... yes
checking for rb_sym2str() in ruby.h... yes
checking for rb_str_intern() in ruby.h... yes
checking for rb_integer_unpack() in ruby.h... yes
creating Makefile

current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/cbor-0.5.9.6/ext/cbor
make DESTDIR\= sitearchdir\=./.gem.20240529-881884-8cxfn4 sitelibdir\=./.gem.20240529-881884-8cxfn4 clean

current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/cbor-0.5.9.6/ext/cbor
make DESTDIR\= sitearchdir\=./.gem.20240529-881884-8cxfn4 sitelibdir\=./.gem.20240529-881884-8cxfn4
compiling buffer.c
In file included from sysdep.h:30,
                 from buffer.h:33,
                 from buffer.c:28:
buffer.c: In function ‘CBOR_buffer_static_init’:
renamer.h:32:36: warning: old-style function definition [-Wold-style-definition]
   32 | #define msgpack_buffer_static_init CBOR_buffer_static_init
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~
buffer.c:46:6: note: in expansion of macro ‘msgpack_buffer_static_init’
   46 | void msgpack_buffer_static_init()
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~
buffer.c: In function ‘CBOR_buffer_static_destroy’:
renamer.h:31:39: warning: old-style function definition [-Wold-style-definition]
   31 | #define msgpack_buffer_static_destroy CBOR_buffer_static_destroy
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~
buffer.c:63:6: note: in expansion of macro ‘msgpack_buffer_static_destroy’
   63 | void msgpack_buffer_static_destroy()
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
compiling buffer_class.c
buffer_class.c: In function ‘read_until_eof’:
buffer_class.c:270:17: error: passing argument 3 of ‘rb_rescue2’ from incompatible pointer type [-Wincompatible-pointer-types]
  270 |                 read_until_eof_error, (VALUE)(void*) args,
      |                 ^~~~~~~~~~~~~~~~~~~~
      |                 |
      |                 VALUE (*)(VALUE) {aka long unsigned int (*)(long unsigned int)}
In file included from /var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby/ruby.h:41,
                 from /var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby.h:38,
                 from compat.h:30,
                 from buffer_class.c:28:
/var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby/internal/iterator.h:388:63: note: expected ‘VALUE (*)(VALUE, 
VALUE)’ {aka ‘long unsigned int (*)(long unsigned int,  long unsigned int)’} but argument is of type ‘VALUE (*)(VALUE)’ {aka ‘long
unsigned int (*)(long unsigned int)’}
  388 | VALUE rb_rescue2(VALUE (*b_proc)(VALUE), VALUE data1, VALUE (*r_proc)(VALUE, VALUE), VALUE data2, ...);
      |                                                       ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make: *** [Makefile:248: buffer_class.o] Error 1

make failed, exit code 2

Gem files will remain installed in /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/cbor-0.5.9.6 for inspection.
Results logged to
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/extensions/x86_64-linux/3.2.0/cbor-0.5.9.6/gem_make.out

  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:119:in `run'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:53:in `block in make'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:45:in `each'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:45:in `make'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/ext_conf_builder.rb:42:in `build'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:187:in `build_extension'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:221:in `block in build_extensions'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:218:in `each'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:218:in `build_extensions'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/installer.rb:846:in `build_extensions'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/rubygems_gem_installer.rb:72:in
`build_extensions'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/rubygems_gem_installer.rb:28:in
`install'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/source/rubygems.rb:198:in `install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/gem_installer.rb:54:in
`install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/parallel_installer.rb:156:in
`do_install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/parallel_installer.rb:147:in
`block in worker_pool'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:62:in `apply_func'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:57:in `block in
process_queue'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:54:in `loop'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:54:in `process_queue'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:90:in `block (2 levels) in
create_threads'

An error occurred while installing cbor (0.5.9.6), and Bundler cannot continue.

In Gemfile:
  webauthn was resolved to 3.0.0, which depends on
    cose was resolved to 1.3.0, which depends on
      cbor

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/posix-spawn-0.3.15/ext
/var/lib/mastodon/.rbenv/versions/3.2.3/bin/ruby extconf.rb
creating Makefile

current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/posix-spawn-0.3.15/ext
make DESTDIR\= sitearchdir\=./.gem.20240529-881884-yctwm sitelibdir\=./.gem.20240529-881884-yctwm clean

current directory: /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/posix-spawn-0.3.15/ext
make DESTDIR\= sitearchdir\=./.gem.20240529-881884-yctwm sitelibdir\=./.gem.20240529-881884-yctwm
compiling posix-spawn.c
posix-spawn.c: In function ‘posixspawn_file_actions_init’:
posix-spawn.c:226:34: error: passing argument 2 of ‘rb_hash_foreach’ from incompatible pointer type [-Wincompatible-pointer-types]
  226 |         rb_hash_foreach(options, posixspawn_file_actions_operations_iter, (VALUE)fops);
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                  |
|                                  int (*)(VALUE,  VALUE,  posix_spawn_file_actions_t *) {aka int (*)(long unsigned int, 
long unsigned int,  posix_spawn_file_actions_t *)}
In file included from /var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby/internal/scan_args.h:37,
                 from /var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby/ruby.h:47,
                 from /var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby.h:38,
                 from posix-spawn.c:14:
/var/lib/mastodon/.rbenv/versions/3.2.3/include/ruby-3.2.0/ruby/internal/intern/hash.h:83:40: note: expected ‘int (*)(VALUE, 
VALUE,  VALUE)’ {aka ‘int (*)(long unsigned int,  long unsigned int,  long unsigned int)’} but argument is of type ‘int (*)(VALUE, 
VALUE,  posix_spawn_file_actions_t *)’ {aka ‘int (*)(long unsigned int,  long unsigned int,  posix_spawn_file_actions_t *)’}
   83 | void rb_hash_foreach(VALUE hash, int (*func)(VALUE key, VALUE val, VALUE arg), VALUE arg);
      |                                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
posix-spawn.c: In function ‘Init_posix_spawn_ext’:
posix-spawn.c:480:1: warning: old-style function definition [-Wold-style-definition]
  480 | Init_posix_spawn_ext()
      | ^~~~~~~~~~~~~~~~~~~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make: *** [Makefile:248: posix-spawn.o] Error 1

make failed, exit code 2

Gem files will remain installed in /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/posix-spawn-0.3.15 for
inspection.
Results logged to
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/extensions/x86_64-linux/3.2.0/posix-spawn-0.3.15/gem_make.out

  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:119:in `run'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:53:in `block in make'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:45:in `each'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:45:in `make'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/ext_conf_builder.rb:42:in `build'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:187:in `build_extension'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:221:in `block in build_extensions'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:218:in `each'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/ext/builder.rb:218:in `build_extensions'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/3.2.0/rubygems/installer.rb:846:in `build_extensions'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/rubygems_gem_installer.rb:72:in
`build_extensions'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/rubygems_gem_installer.rb:28:in
`install'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/source/rubygems.rb:198:in `install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/gem_installer.rb:54:in
`install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/parallel_installer.rb:156:in
`do_install'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/installer/parallel_installer.rb:147:in
`block in worker_pool'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:62:in `apply_func'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:57:in `block in
process_queue'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:54:in `loop'
  /var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:54:in `process_queue'
/var/lib/mastodon/.rbenv/versions/3.2.3/lib/ruby/gems/3.2.0/gems/bundler-2.4.13/lib/bundler/worker.rb:90:in `block (2 levels) in
create_threads'

An error occurred while installing posix-spawn (0.3.15), and Bundler cannot continue.

In Gemfile:
  posix-spawn

**Comment by Preskton** [comment]
@pandoli365 I'm seeing the same thing after doing a system update on arch as well. Since you closed the issue, were you able to resolve it? If so, what did you do?

**Comment by Preskton** [comment]
At this point, I've tried with several different ruby versions, all with the same result: it seems there's a drift between what's happening in [`cbor`'s `buffer_class.c`](https://github.com/cabo/cbor-ruby/blob/4e8b2ec982f41cfcd6568a429add205c8259f470/ext/cbor/buffer_class.c#L270-L272) and the underlying library.

This seems related: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/146840

And further digging: https://github.com/cabo/cbor-ruby/pull/19, seems this was fixed in `cbor v0.5.9.7`.

**Comment by Preskton** [comment]
It looks like underlying compiler changes broke this across several libraries:

- cbor
- posix-spawn
- charlock_holmes

Temporarily had to add the following to the `Gemfile`:

```
# temp hacks while upstream masto and dependencies update pins
gem 'cbor', '~> 0.5.9.8'
gem 'posix-spawn', github: "https://github.com/rtomayko/posix-spawn/pull/93"
```

And comment out the original reference to `posix-spawn`.

Then had to run:

```
bundle config build.charlock_holmes  --with-cxxflags='#'
```

as `charlock_holmes` was running into build issues as well.

After all that, I could get a clean `bundle install`. 

Thank you for your help, @ThisIsMissEm!

**Reply by ClearlyClaire** [reply]
We can update `cbor` and drop `posix-spawn` on our next Mastodon patch release, but `charlock_holmes` is a bit more of an issue.

I tried reproducing this on my end, and I can reproduce the issue but `bundle config build.charlock_holmes  --with-cxxflags='#'` does not seem to help on my end.

**Reply by ClearlyClaire** [reply]
Looking at the `gem_make.out` log, bundler seems to parse the configuration incorrectly and read it as `--with-extraflags=`, treating the `#` as the start of a comment. Not sure how you managed to get a different behavior.

EDIT: Unsetting that config line and manually setting the environment variable instead (`BUNDLE_BUILD__CHARLOCK_HOLMES="--with-cxxflags='#'"`) seems to work

**Reply by Preskton** [reply]
Super strange - just tried the same instructions above and it worked, again. I'll blame something funky in our setup. :)
